segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1325 @@
+            Timer timer = new Timer();
+
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+                // Retrieve metadata and make sure it does not exist.
+                SegmentMetadata oldSegmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+                if (null != oldSegmentMetadata) {
+                    throw new StreamSegmentExistsException(streamSegmentName);
+                }
+
+                // Create a new record.
+                SegmentMetadata newSegmentMetatadata = SegmentMetadata.builder()
+                        .name(streamSegmentName)
+                        .maxRollinglength(rollingPolicy.getMaxLength() == 0 ? SegmentRollingPolicy.NO_ROLLING.getMaxLength() : rollingPolicy.getMaxLength())
+                        .ownerEpoch(this.epoch)
+                        .build();
+
+                newSegmentMetatadata.setActive(true);
+                txn.create(newSegmentMetatadata);
+                // commit.
+                txn.commit();
+
+                val retValue =  SegmentStorageHandle.writeHandle(streamSegmentName);
+                Duration elapsed = timer.getElapsed();
+                log.debug("create - segment={} rollingPolicy={} latency={}.", streamSegmentName, rollingPolicy, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "create", traceId, retValue);
+                return retValue;
+            } catch (StorageMetadataAlreadyExistsException ex) {
+                throw new StreamSegmentExistsException(streamSegmentName, ex);
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            }
+        });
+    }
+
+    /**
+     * Writes the given data to the StreamSegment.
+     *
+     * @param handle  A read-write SegmentHandle that points to a Segment to write to.
+     * @param offset  The offset in the StreamSegment to write data at.
+     * @param data    An InputStream representing the data to write.
+     * @param length  The length of the InputStream.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
+     * it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> BadOffsetException: When the given offset does not match the actual length of the segment in storage.
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for this Segment (it was fenced out).
+     * </ul>
+     * @throws IllegalArgumentException If handle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
```

Comment:
One of the limitations of today tier 2 model comes for a 1 segment stream: there is only 1 thread from the tier 2 that moves the chunks to storage. An important question is: with this new implementation, could we also parallelize movement of chunks for 1 Segment case or things will remain the same in this sense?

Reply:
Yes. That is the intention to parallelized writes and reads as much as possible. But will have to do that incrementally in next couple milestone.